crmprtd/process.py

```diff
@@ -111,21 +123,72 @@ def process(
         )
         raise ValueError("No network name given")
 
+    # Establish a database session
+    engine = create_engine(connection_string)
+    Session = sessionmaker(engine)
+    sesh = Session()
+
+    # For certain networks configured in `data/publication_lags.yaml`:
+    # If start_date is not specified, determine it from latest obs time in database.
+    # This query typically takes on the order of 90 s, but that cost is far lower than
+    # that of trying to insert - one by one - observations that are already present.
+    # A publication lag (days) is subtracted from the latest obs time to determine
+    # the start_date.
+
+    with resource_stream("crmprtd", "data/publication_lags.yaml") as pl_f:
+        publication_lags = yaml.safe_load(pl_f)
+
+    if network in publication_lags and start_date is None:
+        log.info(f"Querying database for latest obs time for network {network}...")
```

Comment:
I find this logic very dangerous and do not like it. It's not uncommon for networks to add observations to their archives in a non-monotonically increasing fashion. Filtering obs that are before the most recent obs_time doesn't give the opportunity to catch updates from the network, and if I'm not mistaken, it will completely break any of our use cases where we infill gaps (say if our `crmprtd` VM goes down overnight and is restarted later).

Reply:
I understand and I was not sure this was a good idea either. I'm glad to have this feedback.

Motivation was basically that it takes (as of this PR) a very long time to process an entire BCH download file, because of the insert strategy/ies presently in use. (Improvements to the insertion method mean this may no longer be an issue.) I knew that just using the latest obs time in the database was naive, even assuming the database was up to date with all previous downloads. The publication lag was intended to backstop this, while still allowing us to process only a subset of the records in the download. I recognize the reasoning here may be flawed.

This only applies when the start time is not specified. This does "break" the previous usage: previously leaving it unspecified meant "process all". Now we would have to specify it explicitly.

This whole explanation makes me regret this approach.